### PR TITLE
added basic cypress e2e setup

### LINKED ISSUE #12: Cypress e2e testing setup inc. accessibility, auth
**As a** developer
**I want** there to be [cypress](https://www.cypress.io/) end-to-end tests that run on PR into main
**So that** I can be confident of change I make is good to be deployed

## Acceptance criteria

- [ ] [maybe] Accessibility with [pa11y](https://github.com/pa11y/pa11y)
- [ ] Runs headless on `make test-e2e`

## Followups

- [ ] e2e authz tests

### PR DESCRIPTION
**Pull request by jhughes982**
## Resolves #12 
- This PR adds in a basic scaffold for running e2e tests using cypress
- I've opened as a draft for now for team visibility, there are some additional files and documentation that I will add
- This setup is intended to allow cypress tests to be run both locally and in a docker setup with the `release` version of the app
- Feedback welcome

### Checklist

- [ ] Documentation has been updated

### CODE REVIEW

**Comment by jhughes982** [comment]
Bringing this out of draft now as it is ready for review.

File: web/cypress.config.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+const { defineConfig } = require("cypress");
+
+module.exports = defineConfig({
+  e2e: {
+    baseUrl: process.env.CYPRESS_baseUrl || "http://localhost:5173",
+    supportFile: "cypress/support/e2e.ts",
+    specPattern: "cypress/e2e/**/*.cy.ts",
+    video: false,
```
Comment (jhughes982) [inline web/cypress.config.ts:8]: We can enable videos if anyone feels it is necessary (e.g. for CI)?

File: web/cypress/e2e/app.cy.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    cy.visit("/");
+
+    cy.contains("ARC portal").should("be.visible");
+    cy.contains("fibble").should("not.exist");
+
+    // run accessibility check
+    cy.injectAxe();
+    cy.checkA11y(undefined, {
+      includedImpacts: ["critical", "serious"],
+    });
```
Comment (jhughes982) [inline web/cypress/e2e/app.cy.ts:14]: There is a lot of customisation possible with the axe library. I made the default check focus on critical and serious accessibility flaws which seems like a reasonable place to start. But I'm open to opinions.

We might need to add more customisation soon as it seemed like the accessibility test failures could have been more descriptive in showing you what actually caused the failure. But let's see how we get on and we can add customisation if needed.

File: web/cypress/fixtures/example.json [hunk 3]
```diff
@@ -0,0 +1,5 @@
+{
+  "name": "Using fixtures to represent data",
+  "email": "hello@cypress.io",
```
Comment (jhughes982) [inline web/cypress/fixtures/example.json:3]: I imagine we will need to add fixtures and intercepts for many of the portal features and to test edge cases. For instance the auth flow might require it.

File: web/cypress/support/commands.ts
Comment (jhughes982) [inline web/cypress/support/commands.ts:1]: This is just boilerplate for now. Custom commands will be necessary as soon as we start adding real tests.

File: web/cypress/support/e2e.ts
Comment (jhughes982) [inline web/cypress/support/e2e.ts:1]: More boilerplate, will be fleshed out when we start adding tests.

File: web/cypress/support/index.d.ts
Comment (jhughes982) [inline web/cypress/support/index.d.ts:1]: Same boilerplate as above...

File: web/eslint.config.js [hunk 4]
```diff
@@ -25,5 +25,13 @@ export default tseslint.config(
...
       "react-refresh/only-export-components": ["warn", { allowConstantExport: true }],
     },
+  },
+  {
+    files: ["cypress.config.ts"],
+    rules: {
+      "@typescript-eslint/no-require-imports": "off",
+      "@typescript-eslint/no-unsafe-assignment": "off",
+      "@typescript-eslint/no-unsafe-call": "off",
+    },
```
Comment (jhughes982) [inline web/eslint.config.js:35]: I got some frustrating typescript and eslint conflicting errors, so I had to add this to allow the cypress config to pass. This should be ok as it is just the standard cypress config that doesn't contain any critical app logic.

File: web/src/index.css
Comment (jhughes982) [inline web/src/index.css:1]: I updated this file so that it passed the basic accessibility checks.

File: web/src/App.tsx
Comment (jhughes982) [inline web/src/App.tsx:1]: File was updated to pass accessibility tests.

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -31,4 +32,12 @@ jobs:
       - uses: actions/checkout@v4
 
       - name: Run e2e tests
-        run: make test-e2e
+        run: make test-e2e-release
+        working-directory: ${{ github.workspace }}
+
+      - name: Upload Cypress screenshots (on failure)
```
Comment (t-young31) [inline .github/workflows/test.yml:38]: this is a smidge scary, as the workflow will execute with secrets and they'd be being uploaded publicly(?)

File: Makefile [hunk 6]
```diff
@@ -28,8 +28,17 @@ test:  ## Run all tests
 test-unit:  ## Run unit tests
 	go test ./...
 
-test-e2e:  ## Run end-to-end tests
-	cd e2e && docker compose build && docker compose run test && docker compose down --remove-orphans
+# Run Cypress locally against dockerised dev server
+# make sure the dockerised dev server is running first with `make dev`
+test-e2e-dev:
+	cd web && CYPRESS_baseUrl=http://localhost:8000 npx cypress run --headless --browser chrome
```
Comment (t-young31) [inline Makefile:34]: code > comments
```suggestion
test-e2e-dev:
	if ! docker compose -p $(DEV_PROJECT_NAME) ps --services --filter "status=running" | grep nginx; then \
		echo "dev environment not running"; exit 1; \
	fi
	cd web && CYPRESS_baseUrl=http://localhost:8000 npx cypress run --headless --browser chrome
```

File: Makefile [hunk 7]
```diff
@@ -28,8 +28,17 @@ test:  ## Run all tests
...
+
+## Run cypress against the release build
+test-e2e-release:
+	cd e2e && \
+	docker compose -p "ucl-arc-tre-e2e" build && \
```
Comment (t-young31) [inline Makefile:39]: probably worth extracting this into a var cf. https://github.com/ucl-arc-tre/portal/blob/cc1f4055164ee528605cddaae9ec4476ce43969c/Makefile#L2

File: e2e/docker-compose.yaml [hunk 8]
```diff
@@ -40,3 +33,21 @@ services:
...
+  cypress:
+    image: cypress/included:14.3.0
+    depends_on:
+      - nginx
+    working_dir: /web
+    volumes:
+      - ../web:/web
+    environment:
+      - CYPRESS_baseUrl=http://nginx:80
+      - CYPRESS_VIDEO=false
```
Comment (t-young31) [inline e2e/docker-compose.yaml:46]: don't need to define it in 2 places(?)
```suggestion
```

File: e2e/docker-compose.yaml [hunk 9]
```diff
@@ -40,3 +33,21 @@ services:
...
+    entrypoint: /bin/sh
+    command: >
+      -c "
+      npm ci &&
+      npx cypress run --config-file cypress.config.ts --browser electron &&
+      npx cypress run --config-file cypress.config.ts --browser firefox
```
Comment (t-young31) [inline e2e/docker-compose.yaml:52]: why is dev chrome and release electron/firefox? chrome+firefox for both sounds sensible?

File: web/README.md [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+This React + Typescript project was set up using [Vite](https://vitejs.dev/guide/why.html).
+
+Development files are located in '/src' with all components routed into App.ts.
+
+---
+
+# Requirements
+
+Make sure you have read through the `README.md` file at the PROJECT_ROOT directory. This will give you more background on the application architecture.
```
Comment (t-young31) [inline web/README.md:11]:
```suggestion
Make sure you have read through the root [`README.md`](../README.md) file. This will give you more background on the application architecture.
```

File: web/README.md [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+If needed, you can also update the version numbers of individual packages manually in the `package.json` file. Running `npm install` again will install the specified dependencies.
+
+To automatically check for package updates you can run:
+
+```javascript
+// add a -u flag at the end of the command to automatically update everything in package.json
+npx npm-check-updates
+```
+
+This will check if any updates are avilable for packages defined in `package.json`. Once your `package.json` has the latest version numbers, just run `npm install` to install them.
```
Comment (t-young31) [inline web/README.md:38]: I'd rm this and just rely on dependabot – feel free to disagree

File: web/cypress.config.ts [hunk 12]
```diff
@@ -0,0 +1,11 @@
+const { defineConfig } = require("cypress");
+
+module.exports = defineConfig({
+  e2e: {
+    baseUrl: process.env.CYPRESS_baseUrl || "http://localhost:5173",
```
Comment (t-young31) [inline web/cypress.config.ts:5]: I'd be tempted to default this to :8000 as we'll soon have tests that hit the API

File: web/cypress/e2e/app.cy.ts [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+
+describe("ARC Portal UI", () => {
+  it("shows the ARC portal phrase on the homepage", () => {
+    cy.visit("/");
+
+    cy.contains("ARC portal").should("be.visible");
+    cy.contains("fibble").should("not.exist");
+
+    // run accessibility check
+    cy.injectAxe();
```
Comment (t-young31) [inline web/cypress/e2e/app.cy.ts:11]: should be a separate test(?)

File: web/cypress/e2e/app.cy.ts (hunk 2 above)
Comment (t-young31) [reply web/cypress/e2e/app.cy.ts:14]: sounds like a reasonable start. running this locally I get an error with `serious` 😞

File: web/cypress/support/commands.ts
Comment (t-young31) [inline web/cypress/support/commands.ts:1]: why's this all commented out?

**Review by t-young31 — COMMENTED** [review]
🙌🏼

File: web/cypress/e2e/app.cy.ts (hunk 2 above)
Comment (t-young31) [reply web/cypress/e2e/app.cy.ts:14]: said error:
Screenshot 2025-04-14 at 10 24 07

Comment (jhughes982) [reply web/cypress/e2e/app.cy.ts]: https://github.com/ucl-arc-tre/portal/pull/33/commits/00310c5459b71c6bf87e8b911a3792e1b91d9d47

This was an example of being caught out by brief flashes of content while the app is loading. The `login` button briefly flashes on load but disappears very quickly if you are already logged in. So the test was passing sometimes but occasionally the test was quick enough to catch the login button before it disappeared.

I updated the css to fix the accessibility issue. Try running it again.

Comment (jhughes982) [reply web/cypress/e2e/app.cy.ts]: Ah, its still failing. I'll have another look.

Comment (t-young31) [reply web/cypress/e2e/app.cy.ts]: > briefly flashes on load but disappears very quickly if you are already logged in

flashiness rm-ed in https://github.com/ucl-arc-tre/portal/pull/32/files#diff-2e5aec0d80c33c218d97cb48edb61724a5e25bfc33478de5907916a90a92accf 😅

Comment (jhughes982) [reply web/cypress/e2e/app.cy.ts]: I think it is now fixed here: https://github.com/ucl-arc-tre/portal/pull/33/commits/7c7adf307944a470b31bd154bcdb6d1a6e11556e

This is interesting. The issue is that the release version of the tests run in light mode whereas other environments run in dark mode (depending on your local settings probably). It just needed a bit of css adding to specifically target light mode to make the login button visible.

Comment (jhughes982) [reply web/cypress/e2e/app.cy.ts]: That might warrant a bit of documentation to explain that, unless we want to make the tests more dynamic to run across both light and dark mode. @acholyn have you done that before?

Comment (jhughes982) [reply web/cypress/e2e/app.cy.ts]: We could do something like this:

```
describe('Light mode test', () => {
  it('should render in light mode', () => {
    cy.setColorScheme('light');
    
    // run tests
  });
});
```

Did we decide we were going to support both light and dark mode in the end?

File: .github/workflows/test.yml (hunk 5 above)
Comment (jhughes982) [reply .github/workflows/test.yml:38]: We can remove the screenshots feature if you prefer?

File: web/cypress/e2e/app.cy.ts (hunk 13 above)
Comment (jhughes982) [reply web/cypress/e2e/app.cy.ts]: I've now extracted it into a separate test. The exact testing approach is yet to be decided though. For instance, shall we have a separate testing file dedicated to running a full accessibility testing suite?

File: e2e/docker-compose.yaml (hunk 9 above)
Comment (jhughes982) [reply e2e/docker-compose.yaml:52]: Electron is a stripped down version of chrome, it achieves the same thing and it means we don't have to install chrome within the workflow. So it cuts down on action minutes and makes the setup less bloated. Happy to reconsider though if you feel strongly?

File: web/cypress/e2e/app.cy.ts (hunk 2 above)
Comment (t-young31) [reply web/cypress/e2e/app.cy.ts]: I think light+dark mode accessibility tests would be the way to go, if indeed we're supporting both, which we probably are(?)

File: .github/workflows/test.yml (hunk 5 above)
Comment (t-young31) [reply .github/workflows/test.yml:38]: I think I would, but needs to be weighed up by how useful they are. inclined to say if you need a screenshot then run it locally

File: e2e/docker-compose.yaml (hunk 9 above)
Comment (t-young31) [reply e2e/docker-compose.yaml:52]: I think it depends how similar electron and chrome are – if there's almost no difference then very happy to leave as is

File: web/cypress/support/commands.ts
Comment (t-young31) [reply web/cypress/support/commands.ts:1]: 👀 https://github.com/ucl-arc-tre/portal/pull/33#discussion_r2039761991

File: e2e/docker-compose.yaml (hunk 9 above)
Comment (jhughes982) [reply e2e/docker-compose.yaml:52]: They both run on chromium and both have full access to the HTML/CSS/JS rendered content. So I'd be inclined to leave this as a simpler setup. If we start to get behaviour differences we can always add chrome on as an extension if needed later.

File: .github/workflows/test.yml (hunk 5 above)
Comment (jhughes982) [reply .github/workflows/test.yml:38]: Sounds reasonable to me. Another option could be to do something like this:

```
    volumes:
      - ../web:/web
      - ../web/cypress/screenshots:/web/cypress/screenshots
```

So that the screenshots for the release build are available locally, and during CI they don't get uploaded as artefacts. What do you reckon?

File: web/cypress.config.ts (hunk 12 above)
Comment (jhughes982) [reply web/cypress.config.ts]: https://github.com/ucl-arc-tre/portal/pull/33/commits/27ac4d700343daf874602455e0bfa00ade24a20d

The tests now fall back to `localhost:8000` . The readme instructions have also been simplified to reflect this.

File: Makefile (hunk 7 above)
Comment (jhughes982) [reply Makefile]: https://github.com/ucl-arc-tre/portal/pull/33/commits/d99a7b070c491d43a0ce9b8b9263f40f543e3398
